src/Bicep.Core/Semantics/Namespaces/DefaultNamespaceProvider.cs

```diff
@@ -14,60 +16,60 @@ namespace Bicep.Core.Semantics.Namespaces;
-
-        this.azResourceTypeLoaderFactory = azResourceTypeLoaderFactory;
+        this.resourceTypeLoaderFactory = resourceTypeLoaderFactory;
         this.providerLookup = new Dictionary<string, GetNamespaceDelegate>
         {
-            [SystemNamespaceType.BuiltInName] = (alias, scope, features, sourceFileKind, version) => SystemNamespaceType.Create(alias, features, sourceFileKind),
-            [AzNamespaceType.BuiltInName] = (alias, scope, features, sourceFileKind, version) =>
-            {
-                AzResourceTypeProvider? provider = builtInAzResourceTypeProvider;
-                if (features.DynamicTypeLoadingEnabled && version is not null)
-                {
-                    // TODO(asilverman): Current implementation of dynamic type loading needs to be refactored to handle caching of the provider to optimize 
-                    // performance hit as a result of recreating the provider for each call to TryGetNamespace.
-                    // Tracked by - https://msazure.visualstudio.com/One/_sprints/taskboard/Azure-ARM-Deployments/One/Custom/Azure-ARM/Gallium/Jul-2023?workitem=24563440
-                    var loader = azResourceTypeLoaderFactory.GetResourceTypeLoader(version, features);
-                    if (loader is null)
-                    {
-                        return null;
-                    }
-                    var overriddenProviderVersion = builtInAzResourceTypeLoaderVersion;
-                    if (features.DynamicTypeLoadingEnabled)
-                    {
-                        overriddenProviderVersion = version ?? overriddenProviderVersion;
-                    }
-                    provider = new AzResourceTypeProvider(loader, overriddenProviderVersion);
-                }
-                return AzNamespaceType.Create(alias, scope, provider, sourceFileKind);
-            },
-            [K8sNamespaceType.BuiltInName] = (alias, scope, features, sourceFileKind, version) => K8sNamespaceType.Create(alias),
-            [MicrosoftGraphNamespaceType.BuiltInName] = (alias, scope, features, sourceFileKind, version) => MicrosoftGraphNamespaceType.Create(alias),
+            [AzNamespaceType.BuiltInName] = CreateAzNamespace,
+            [SystemNamespaceType.BuiltInName] = (providerDescriptor, resourceScope, features, sourceFileKind) => SystemNamespaceType.Create(providerDescriptor.Alias, features, sourceFileKind),
+            [K8sNamespaceType.BuiltInName] = (providerDescriptor, resourceScope, features, sourceFileKind) => K8sNamespaceType.Create(providerDescriptor.Alias),
+            [MicrosoftGraphNamespaceType.BuiltInName] = (providerDescriptor, resourceScope, features, sourceFileKind) => MicrosoftGraphNamespaceType.Create(providerDescriptor.Alias),
         }.ToImmutableDictionary();
     }
 
+    private NamespaceType? CreateAzNamespace(ResourceTypesProviderDescriptor providerDescriptor, ResourceScope scope, IFeatureProvider features, BicepSourceFileKind sourceFileKind)
```

Comment:
[nit] The convention we've been following for functions that can return null is to prefix the name with `Try` - e.g. `TryCreateAzNamespace`